atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssignerTest.java

```diff
@@ -0,0 +1,474 @@
+        when(bucketWriter.writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15)))
+                .thenReturn(WriteState.FAILURE);
+
+        bucketAssigner.run();
+        assertBucketState(BucketState.CONTENTION_ON_WRITES);
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(1, BucketAssignerState.immediatelyClosing(1, 15)));
+    }
+
+    @Test
+    public void startTransitionsStillTransitionsToImmediatelyClosingEvenIfBucketWritesAtImmediatelyClosingThrows() {
+        stateTable.setInitialState(1, BucketAssignerState.start(1));
+        setPunchTable(1, 15);
+        clock.advance(Duration.ofMinutes(10));
+        when(bucketWriter.writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15)))
+                .thenThrow(RUNTIME_EXCEPTION);
+
+        assertThatThrownBy(bucketAssigner::run).isEqualTo(RUNTIME_EXCEPTION);
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(1, BucketAssignerState.immediatelyClosing(1, 15)));
+    }
+
+    @Test
+    public void immediatelyClosingStaysInImmediatelyClosingIfBucketWritesFail() {
+        stateTable.setInitialState(1, BucketAssignerState.immediatelyClosing(1, 15));
+        setPunchTable(1, 15);
+        when(bucketWriter.writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15)))
+                .thenReturn(WriteState.FAILURE);
+
+        bucketAssigner.run();
+        assertBucketState(BucketState.CONTENTION_ON_WRITES);
+        stateTable.assertThatStateMachineTransitionedThroughTo(); // No transition
+    }
+
+    @Test
+    public void immediatelyClosingUsesOriginalTimestampsEvenIfPunchTableHasUpdated() {
+        stateTable.setInitialState(1, BucketAssignerState.immediatelyClosing(1, 15));
+        setPunchTable(2, 20);
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(10)), 15, 25);
+        bucketAssigner.run();
+
+        assertBucketState(BucketState.INCOMPLETE);
+        verify(bucketWriter).writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15));
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(2, BucketAssignerState.start(15)),
+                BucketStateAndIdentifier.of(2, BucketAssignerState.opening(15)),
+                BucketStateAndIdentifier.of(2, BucketAssignerState.waitingUntilCloseable(15)));
+    }
+
+    @Test
+    public void immediatelyClosingBubblesUpBucketWriteExceptionWithNoTransition() {
+        stateTable.setInitialState(1, BucketAssignerState.immediatelyClosing(1, 15));
+        setPunchTable(1, 15);
+
+        when(bucketWriter.writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15)))
+                .thenThrow(RUNTIME_EXCEPTION);
+
+        assertThatThrownBy(bucketAssigner::run).isEqualTo(RUNTIME_EXCEPTION);
+        stateTable.assertThatStateMachineTransitionedThroughTo(); // No transition
+    }
```

Comment:
not actionable: There seems to have to be some way to reduce the duplication here; I'm not opposed to this given the criticality of the code under test, but it seems like we have a lot of "bubbles up exception" and "uses original timestamps even if the punch table has changed" kind of tests